src/core/agent/services/multiSigService.ts

```diff
@@ -744,6 +752,168 @@ class MultiSigService extends AgentService {
+      );
+      const atc = roleims.substring(rpy.size);
+      const roleEmbeds = {
+        rpy: [rpy, atc],
+      };
+
+      await this.sendMultisigExn(
+        ourIdentifier?.signifyName,
+        ourAid,
+        MultiSigRoute.RPY,
+        roleEmbeds,
+        recp,
+        { gid: aid }
+      );
+      results.push(op);
+    }
+    return results;
+  }
+
+  async joinAuthorization(notificationSaid: string) {
+    const exchangeMessage = await this.props.signifyClient
+      .exchanges()
+      .get(notificationSaid);
+    if (!exchangeMessage) {
+      throw new Error(
+        `${MultiSigService.EXN_MESSAGE_NOT_FOUND} ${notificationSaid}`
+      );
+    }
+    const multisigAid = exchangeMessage.exn.a.gid;
+    const multisigMetadataRecord =
+      await this.identifierStorage.getIdentifierMetadata(multisigAid);
+    const multisigSignifyName = multisigMetadataRecord.signifyName;
+    const request = await this.props.signifyClient
+      .groups()
+      .getRequest(notificationSaid)
+      .catch((error) => {
+        const errorStack = (error as Error).stack as string;
+        const status = errorStack.split("-")[1];
+        if (/404/gi.test(status) && /SignifyClient/gi.test(errorStack)) {
+          return [];
+        } else {
+          throw error;
+        }
+      });
+    if (!request.length) {
+      throw new Error(
+        `${MultiSigService.GROUP_REQUEST_NOT_FOUND} ${notificationSaid}`
+      );
+    }
+    const exn = request[0].exn;
+    // stamp, eid and role are provided in the exn message
+    const rpystamp = exn.e.rpy.dt;
+    const rpyrole = exn.e.rpy.a.role;
+    const rpyeid = exn.e.rpy.a.eid;
+    const endRoleRes = await this.props.signifyClient
+      .identifiers()
+      .addEndRole(multisigSignifyName, rpyrole, rpyeid, rpystamp);
+    const op = await endRoleRes.op();
+    const rpy = endRoleRes.serder;
+    const sigs = endRoleRes.sigs;
+
+    const hab = await this.props.signifyClient
+      .identifiers()
+      .get(multisigSignifyName);
+    const mstate = hab["state"];
+    const seal = [
+      "SealEvent",
+      { i: hab["prefix"], s: mstate["ee"]["s"], d: mstate["ee"]["d"] },
+    ];
+    const sigers = sigs.map((sig) => new Siger({ qb64: sig }));
+    const roleims = d(
+      messagize(rpy, sigers, seal, undefined, undefined, false)
+    );
+    const atc = roleims.substring(rpy.size);
+    const roleEmbeds = {
+      rpy: [rpy, atc],
+    };
+    const { ourIdentifier, multisigMembers } =
+      await this.getMultisigParticipants(multisigSignifyName);
+    const recp = multisigMembers
+      .filter((signing: any) => signing.aid !== ourIdentifier.id)
+      .map((member: any) => member.aid);
+    const ourAid: Aid = await this.props.signifyClient
+      .identifiers()
+      .get(ourIdentifier.signifyName as string);
+
+    await this.sendMultisigExn(
+      ourIdentifier.signifyName,
+      ourAid,
+      MultiSigRoute.RPY,
```

Comment:
If we are sending this back - won't for example Alice get rpy messages back from Bob and Carol when they join? And then try to join again...

I think we need a way to just process the new end role authorisations, right?

Reply:
I tested this many times, they need to send this back in order to complete the `op`. When Alice tries to join again and send Bob and Carol the exchange, they won't receive those messages. We can check for the old authorization by checking the response from the HTTP get of the OOBI url. What do you think?

Reply:
Or we can add authorized eids into our storage metadata to check it later.

Reply:
Yep, for sure they need to send it back - otherwise Alice won't receive the signatures she needs to add to her OOBI GET response.

> We can check for the old authorization by checking the response from the HTTP get of the OOBI url.

They will only appear once a threshold amount accepts, so we can still get duplicates.

> When Alice tries to join again and send Bob and Carol the exchange, they won't receive those messages.

OK, seems KERIA stops the duplication. Still some wasted effort. We can store the authorised eids locally, this would work but maybe it's OK to ignore it if it stops.

Just thinking there, for a 3 of 3 multi-sig - both Alice and Bob will have a pending operation for the long running operation. So both will call `endRoleAuthorization` at the same time potentially, right?

Just thinking too - we set it so `endRoleAuthorization` runs when the operation is complete. If it's 1 of 3 multi-sig, it will be complete straight away. I think we need to also call `endRoleAuthorization` when creating or joining the multi-sig if the operation is already complete.

Can you test these scenarios, and make sure that the messages stop as you say?

Reply:
> Just thinking there, for a 3 of 3 multi-sig - both Alice and Bob will have a pending operation for the long running operation. So both will call endRoleAuthorization at the same time potentially, right?

We only need one person to call it, I've updated it so only the initiator call this, then others just need to join.

> Just thinking too - we set it so endRoleAuthorization runs when the operation is complete. If it's 1 of 3 multi-sig, it will be complete straight away. I think we need to also call endRoleAuthorization when creating or joining the multi-sig if the operation is already complete.

We still have operation tracking here, right? If so it would still trigger the authorization.

Reply:
> We only need one person to call it, I've updated it so only the initiator call this, then others just need to join.

Hmm, but if the messages eventually stop repeating, then they can all send it, right? Just in case the initiator is offline. I mean technically - if all 3 add the end role and send the messages - then no one needs to join, right?

> We still have operation tracking here, right? If so it would still trigger the authorization.

If the operation is done immediately, no operation tracking - it's only if its still pending

Reply:
Actually, I would keep the join in case we need to be interoperable with other wallets who don't automatically do the end role authorisation.

Can you test this without your check for `groupInitiator` - so all wallets call `endRoleAuthorization` - try with a 1 of 3 multi-sig so they all do it immediately (please implement this, as I said, no long running operation in this case).

In this scenario - it will be "complete" straight away from anyone signing - each wallet will send 3 messages to each other wallet. Upon receiving rpy messages, we will send more from Signify, _but_ the other wallets will receive nothing since the authorisation is complete for the `eid`s - is this correct? Important that we fully understand all the details :)

Reply:
> If the operation is done immediately, no operation tracking - it's only if its still pending

I just tested. The authorization still triggers in this case even though the operation is done right away. I still see the `onSignifyOperationStateChanged` get called :D

> Hmm, but if the messages eventually stop repeating, then they can all send it, right? Just in case the initiator is offline. I mean technically - if all 3 add the end role and send the messages - then no one needs to join, right?

They can all send it. However this is not right "if all 3 add the end role and send the messages - then no one needs to join", other members still need to join the request in order to complete the operation, else the oobi will never appear. 

> each wallet will send 3 messages to each other wallet. Upon receiving rpy messages, we will send more from Signify, but the other wallets will receive nothing since the authorisation is complete for the eids - is this correct? Important that we fully understand all the details :)

It is not really like that, as I am seeing, we will receives 9 rpy notifications on each member. 6 notifications from 2 other members are understandable but looks like it still tries to join the authorization one more time before it stops receiving the rpy notifications and we got 3 more notifications on each side.

Reply:
Can you show me where the multi-sig operation is created if it is not pending? Because in these lines https://github.com/cardano-foundation/cf-identity-wallet/blob/feat/create-multi-sig-oobi/src/core/agent/services/multiSigService.ts#L133-L141 it will not be created. `onSignifyOperationStateChanged` only operates on operations from the DB.

> They can all send it. However this is not right "if all 3 add the end role and send the messages - then no one needs to join", other members still need to join the request in order to complete the operation, else the oobi will never appear.

OK.

> It is not really like that, as I am seeing, we will receives 9 rpy notifications on each member. 6 notifications from 2 other members are understandable but looks like it still tries to join the authorization one more time before it stops receiving the rpy notifications and we got 3 more notifications on each side.

Alright, but it does eventually stop. The problem with only the initiator adding the end role is we may create the multi-sig with a different wallet from another company, and they might not add the end role authorisation. I think it's still better for all to kick it off.

But sure, since it's now 9, which will grow if there's >3 in the multi-sig - perhaps we can add this `authorized_eids` field to the DB to check which we have called join for to optimise?

Reply:
> Can you show me where the multi-sig operation is created if it is not pending? Because in these lines https://github.com/cardano-foundation/cf-identity-wallet/blob/feat/create-multi-sig-oobi/src/core/agent/services/multiSigService.ts#L133-L141 it will not be created. onSignifyOperationStateChanged only operates on operations from the DB.

It still needs to wait for other parties to join before the op is done.

Reply:
I thought in the case of a "1 of 2" multi-sig, (with no witnesses), the operation would be done because a threshold number of signing keys have joined?

Reply:
> But sure, since it's now 9, which will grow if there's >3 in the multi-sig - perhaps we can add this authorized_eids field to the DB to check which we have called join for to optimise?

Ok. I am planning to put the field into the multisig metadata. Is it ok? Or you want to put it somewhere else?

Reply:
> I thought in the case of a "1 of 2" multi-sig, (with no witnesses), the operation would be done because a threshold number of signing keys have joined?

I tested with "1 of 3" multi-sig this morning and the `onSignifyOperationStateChanged` still got called

Reply:
Yeah, I think this works. So `groupMetadata` for the member identifier that the user is using in this wallet. Since it's signed by that key.